akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -510,9 +517,28 @@ object ShardRegion {
     var remaining = entities
 
     def receive: Receive = {
-      case ReceiveTimeout =>
+      case Terminated(ref) =>
+        remaining -= ref
+        if (remaining.isEmpty) {
+          replyTo ! ShardStopped(shard)
+          context.stop(self)
+        }
+
+      case StopTimeoutWarning =>
+        log.warning(
+          s"$typeName: [${remaining.size}] of the entities in shard [{}] not stopped after [{}]. " +
+          "Maybe the handOffStopMessage [{}] is not handled? {}",
+          shard,
+          StopTimeoutWarningAfter.toCoarsest,
+          stopMessage.getClass.getName,
+          if (CoordinatedShutdown(context.system).getShutdownReason().isPresent)
+            "" // the region will be shutdown earlier so would be confusing to say more
+          else
+            s"Waiting additional [${handoffTimeout.toCoarsest}] before stopping the remaining entities.")
```

Comment:
Mix of interpolation and `{}`. Is that because we reached the limit of the logging arity? Don't we have an `Array` variant?

Reply:
That's right. I don't think wrapping in array will make it better. It's on warning level so it will "always" be enabled anyway.